components/nimbus/src/tests/stateful/helpers.rs

```diff
@@ -0,0 +1,98 @@
+    metrics::{EnrollmentStatusExtraDef, MetricsHandler},
+};
+use glean::RecordedEvent;
+use std::sync::{Mutex, MutexGuard};
+
+impl From<EnrollmentStatusExtraDef> for EnrollmentStatusExtra {
+    fn from(enrollment_status_extra: EnrollmentStatusExtraDef) -> Self {
+        EnrollmentStatusExtra {
+            branch: enrollment_status_extra.branch,
+            conflict_slug: enrollment_status_extra.conflict_slug,
+            error_string: enrollment_status_extra.error_string,
+            reason: enrollment_status_extra.reason,
+            slug: enrollment_status_extra.slug,
+            status: enrollment_status_extra.status,
+        }
+    }
+}
+
+/// Implemented on RecordedEvents to aid in verifying the value of keys on event extras
+pub trait GetExtraKey {
+    fn get_extra_key(&self, key: &str) -> String;
+}
+
+impl GetExtraKey for RecordedEvent {
+    fn get_extra_key(&self, key: &str) -> String {
+        self.extra.clone().unwrap().get(key).unwrap().to_owned()
+    }
+}
+
+static METRICS_SINGLE_THREAD_MUTEX: Mutex<()> = Mutex::new(());
+
+/// A Rust implementation of the MetricsHandler trait
+/// Used to test recording of Glean metrics across the FFI within Rust
+///
+/// *NOTE: Use this struct's `new` method when instantiating it to lock the Glean store*
+pub struct TestMetrics<'a> {
+    _lock: Option<MutexGuard<'a, ()>>,
+}
+
+impl<'a> TestMetrics<'a> {
+    /// This struct does two things on initialization:
+    /// 1. It resets Glean
+    /// 2. It locks a static Mutex and hold onto its guard
+    ///
+    /// The purpose of the latter is so it can be used as such at the beginning of a test:
+    /// ```rust
+    /// let metrics = TestMetrics::new();
+    /// ```
+    /// and it will lock the Mutex so no other tests that use Glean can run at the same time.
+    ///
+    /// *NOTE: Cloned copies of TestMetrics objects do not lock the Mutex*
+    pub fn new() -> Self {
+        let lock = match METRICS_SINGLE_THREAD_MUTEX.lock() {
+            Ok(l) => l,
+            Err(poisoned) => poisoned.into_inner(),
+        };
+        let dir = tempfile::tempdir().unwrap();
+
+        let cfg = glean::Configuration {
+            data_path: dir.path().clone().to_path_buf(),
+            application_id: "app-id".into(),
+            upload_enabled: true,
+            server_endpoint: Some("invalid-test-host".into()),
+            max_events: None,
+            delay_ping_lifetime_io: false,
+            uploader: None,
+            use_core_mps: false,
+            trim_data_to_registered_pings: false,
+            log_level: None,
+            rate_limit: None,
+            enable_event_timestamps: false,
+        };
+        let client_info = glean::ClientInfoMetrics::unknown();
+        glean::test_reset_glean(cfg, client_info, true);
+        Self { _lock: Some(lock) }
+    }
+}
+
+impl<'a> Clone for TestMetrics<'a> {
+    fn clone(&self) -> Self {
+        Self { _lock: None }
+    }
+}
+
+unsafe impl<'a> Send for TestMetrics<'a> {}
```

Comment:
This, however, does feel cursed. But it's just for testing and it's documented in a comment so ehhhhhhh 🤷